File: tests/fixtures/python_repo/src/app/services/user_service.py
"""UserService — business logic layer for user management."""

from sqlalchemy.orm import Session

from app.models.user import User
from app.repositories.user_repository import UserRepository
from app.schemas.user import UserCreate, UserUpdate
from app.services.auth_service import hash_password


class UserService:
    def __init__(self, db: Session) -> None:
        self._repo = UserRepository(db)

    def create_user(self, data: UserCreate) -> User:
        """Create a new user, hashing the password before persistence."""
        if self._repo.get_by_email(data.email):
            raise ValueError(f"Email {data.email!r} is already registered")
        if self._repo.get_by_username(data.username):
            raise ValueError(f"Username {data.username!r} is already taken")
        hashed = hash_password(data.password)
        return self._repo.create(email=data.email, username=data.username, hashed_password=hashed)

    def get_user_by_id(self, user_id: int) -> User | None:
        """Return user by primary key, or None."""
        return self._repo.get_by_id(user_id)

    def get_user_by_email(self, email: str) -> User | None:
        return self._repo.get_by_email(email)

    def list_users(self, skip: int = 0, limit: int = 20) -> list[User]:
        return self._repo.list_all(skip=skip, limit=limit)

    def update_user(self, user_id: int, data: UserUpdate) -> User | None:
        """Partially update user fields; re-hashes password if provided."""
        user = self._repo.get_by_id(user_id)
        if user is None:
            return None
        if data.password is not None:
            data = data.model_copy(update={"password": None})
            self._repo.set_password(user, hash_password(data.password or ""))
        return self._repo.update(user, data)

    def delete_user(self, user_id: int) -> bool:
        """Delete user by id; return True if found and deleted."""
        user = self._repo.get_by_id(user_id)
        if user is None:
            return False
        self._repo.delete(user)
        return True

File: tests/fixtures/ts_repo/src/services/userService.ts
/**
 * services/userService.ts — Business logic for user account management.
 */

import { randomUUID } from "crypto";
import type { User, NewUser, UserUpdate, PublicUser } from "../types/user";
import { hashPassword } from "./authService";

// In-memory store; replace with a real DB layer in production
const users: Map<string, User> = new Map();
const byEmail: Map<string, string> = new Map(); // email -> id

export async function createUser(data: NewUser): Promise<PublicUser> {
  if (byEmail.has(data.email)) {
    throw Object.assign(new Error("Email already registered"), { status: 409 });
  }
  const id = randomUUID();
  const passwordHash = await hashPassword(data.password);
  const now = new Date();
  const user: User = {
    id,
    email: data.email,
    username: data.username,
    passwordHash,
    createdAt: now,
    updatedAt: now,
  };
  users.set(id, user);
  byEmail.set(data.email, id);
  return toPublic(user);
}

export function getUserById(id: string): PublicUser {
  const user = users.get(id);
  if (!user) throw Object.assign(new Error("User not found"), { status: 404 });
  return toPublic(user);
}

export function getUserByEmail(email: string): User | undefined {
  const id = byEmail.get(email);
  return id ? users.get(id) : undefined;
}

export async function updateUser(id: string, data: UserUpdate): Promise<PublicUser> {
  const user = users.get(id);
  if (!user) throw Object.assign(new Error("User not found"), { status: 404 });
  if (data.email) {
    byEmail.delete(user.email);
    user.email = data.email;
    byEmail.set(data.email, id);
  }
  if (data.username) user.username = data.username;
  if (data.password) user.passwordHash = await hashPassword(data.password);
  user.updatedAt = new Date();
  users.set(id, user);
  return toPublic(user);
}

export function deleteUser(id: string): void {
  const user = users.get(id);
  if (!user) throw Object.assign(new Error("User not found"), { status: 404 });
  byEmail.delete(user.email);
  users.delete(id);
}

export function listUsers(): PublicUser[] {
  return Array.from(users.values()).map(toPublic);
}

function toPublic(user: User): PublicUser {
  const { id, email, username, createdAt } = user;
  return { id, email, username, createdAt };
}
